File: javascript/Sieve/sieve.js
//NOTE: For this implementation I tried to create my own implementation of The Sieve of Eratosthenes after little research.
// Although not fast enough and more brute force this solution would work.

class Sieve {
  NthPrime(n) {
    let primes = [2, 3, 5, 7];
    // Early return for first 4 primes
    if (n < primes.length) return primes[n];

    let number = 9;

    while (primes.length !== n + 1) {
      // Check division of the primes
      if (!this.CheckDivOfPrimes(number, primes)) {
        primes.push(number);
      }

      // Only check odds as evens are never prime after 2
      number += 2;
    }

    return primes[n];
  }

  CheckDivOfPrimes(n, primes) {
    // Only need to check up to the square root of n
    let limit = Math.ceil(Math.sqrt(n));
    for (let i = 0; i < limit; i++) {
      if (n % primes[i] === 0) {
        return true;
      }
    }

    return false;
  }
}

module.exports = Sieve;

File: go/pkg/sieve/sieve.go
//NOTE: For this implementation I researched The Sieve of Eratosthenes to greater detial and tried to stay as faithful to it as I could

package sieve

type EtSieve struct {
	primes     []int64
	currentMax int64
}

type Sieve interface {
	NthPrime(n int64) int64
}

func NewSieve() Sieve {
	return &EtSieve{
		primes:     []int64{2, 3, 5, 7},
		currentMax: 7,
	}
}

func (s *EtSieve) NthPrime(n int64) int64 {
	// return early for first 4 primes
	if int64(len(s.primes)) > n {
		return s.primes[n]
	}

	// will be used for generating slice length
	limit := s.currentMax

	for int64(len(s.primes)) <= n {
		// Increase the range of the slice
		limit *= 2

		// Create a sieve slice up to the new limit
		sieveArray := make([]bool, limit+1)

		// Mark non-primes in the range only up to the square root of the limit
		for i := int64(2); i*i <= limit; i++ {
			if !sieveArray[i] {
				for j := i * i; j <= limit; j += i {
					sieveArray[j] = true
				}
			}
		}

		// Collect primes from the sieve array
		for i := s.currentMax + 1; i <= limit; i++ {
			if i >= 2 && !sieveArray[i] {
				s.primes = append(s.primes, i)
			}
		}

		s.currentMax = limit
	}

	return s.primes[n]
}
